### PR TITLE
Peer review ui 2 [test ui]

### PR DESCRIPTION
**Pull request by mehalshah**
UI for peer review, but without more react classes

Different enough to justify a new PR

### CODE REVIEW

**Comment by mehalshah** [comment]
image

File: code-studio/src/js/components/progress/course_progress.jsx [hunk 1]
```diff
@@ -24,11 +24,12 @@ const CourseProgress = React.createClass({
   propTypes: {
     professionalLearningCourse: React.PropTypes.bool,
     focusAreaPositions: React.PropTypes.arrayOf(React.PropTypes.number),
-    stages: React.PropTypes.arrayOf(stageShape)
+    stages: React.PropTypes.arrayOf(stageShape),
```
Comment (joshlory) [inline code-studio/src/js/components/progress/course_progress.jsx]: Nit: we've been avoiding trailing commas, but I'm not sure there's a good reason.  FWIW https://github.com/airbnb/javascript#commas recommends trailing commas everywhere.

File: code-studio/src/js/components/progress/progress_dot.jsx [hunk 2]
```diff
@@ -129,6 +134,7 @@ const styles = {
   }
 };
 
+
```
Comment (joshlory) [inline code-studio/src/js/components/progress/progress_dot.jsx]: Nit: extra line.

File: code-studio/src/js/components/progress/progress_dot.jsx [hunk 3]
```diff
@@ -156,49 +162,73 @@ export const ProgressDot = React.createClass({
     const outlineCurrent = this.props.courseOverviewPage && onCurrent;
     const smallDot = !this.props.courseOverviewPage && !onCurrent;
     const showLevelName = level.kind === 'named_level' && this.props.courseOverviewPage;
+    const isPeerReview = level.kind === 'peer_review';
+    const isLockedReview = isPeerReview && level.status === 'not_tried';
```
Comment (joshlory) [inline code-studio/src/js/components/progress/progress_dot.jsx]: We had discussed `locked` being a property on the puzzle dot, not something specific to peer review.

File: code-studio/src/js/components/progress/progress_dot.jsx [hunk 4]
```diff
@@ -156,49 +162,73 @@ export const ProgressDot = React.createClass({
...
+          key='link'
+          href={level.url}
+          onClick={this.props.saveAnswersBeforeNavigation && dotClicked.bind(null, level.url)}
+          style={[styles.outer, (showLevelName || isPeerReview) && {display: 'table-row'}]}
+        >
+          {iconElement}
+          {rowText}
+        </a>
+      );
+    }
```
Comment (joshlory) [inline code-studio/src/js/components/progress/progress_dot.jsx]: [Same feedback as before](https://github.com/code-dot-org/code-dot-org/pull/9221#discussion_r68854287): prefer putting all the JSX inside the return (...) of render() rather than using local variables.

File: code-studio/src/js/components/progress/progress_dot.jsx [hunk 5]
```diff
@@ -156,49 +162,73 @@ export const ProgressDot = React.createClass({
...
   }
 });
-export default connect(state => ({
+export default connect((state, ownProps) => ({
```
Comment (joshlory) [inline code-studio/src/js/components/progress/progress_dot.jsx]: Not needed?

File: code-studio/src/js/progress.js [hunk 6]
```diff
@@ -157,14 +158,33 @@ function loadProgress(scriptData, currentLevelId, saveAnswersBeforeNavigation =
         newProgress[key] = clientState.mergeActivityResult(state.progress[key], action.progress[key]);
       });
 
-      return Object.assign({}, state, {
+      let stages = state.stages.map(stage => Object.assign({}, stage, {levels: stage.levels.map(level => {
```
Comment (joshlory) [inline code-studio/src/js/progress.js]: Can be `const`.  (A few more below.)

File: dashboard/app/models/peer_review.rb [hunk 7]
```diff
@@ -133,4 +139,14 @@ def self.create_for_submission(user_level, level_source_id)
...
   end
+
+  def summarize
+    return Hash.new.tap do |summary_hash|
+      summary_hash[:id] = id
+      summary_hash[:status] = status.nil? ? 'attempted' : 'perfect'
+      summary_hash[:name] = status.nil? ? I18n.t('peer_review.review_in_progress') : I18n.t('peer_review.link_to_submitted_review')
+      summary_hash[:result] = status.nil? ? ActivityConstants::UNSUBMITTED_RESULT : ActivityConstants::BEST_PASS_RESULT
+      summary_hash[:icon] = status.nil? ? '' : 'fa-check'
+    end
```
Comment (joshlory) [inline dashboard/app/models/peer_review.rb]:
```ruby
{
  id: id,
  status: status.nil? ? 'attempted' : 'perfect',
  name: status.nil? ? I18n.t('peer_review.review_in_progress') : I18n.t('peer_review.link_to_submitted_review'),
  result: status.nil? ? ActivityConstants::UNSUBMITTED_RESULT : ActivityConstants::BEST_PASS_RESULT,
  icon: status.nil? ? '' : 'fa-check'
}
```

File: dashboard/app/models/peer_review.rb [hunk 8]
```diff
@@ -133,4 +139,14 @@ def self.create_for_submission(user_level, level_source_id)
...
+  end
```
Comment (joshlory) [inline dashboard/app/models/peer_review.rb]:
``` ruby
def summarize
  {
    id: id,
    status: status.nil? ? 'attempted' : 'perfect',
    name: status.nil? ? I18n.t('peer_review.review_in_progress') : I18n.t('peer_review.link_to_submitted_review'),
    result: status.nil? ? ActivityConstants::UNSUBMITTED_RESULT : ActivityConstants::BEST_PASS_RESULT,
    icon: status.nil? ? '' : 'fa-check'
  }
end
```

File: dashboard/app/models/script.rb [hunk 9]
```diff
@@ -628,4 +644,14 @@ def self.clear_cache
...
+
+  def self.build_property_hash(script_data)
+    Hash.new.tap do |property_hash|
+      property_hash[:pd] = script_data[:pd] || false # default false
+      property_hash[:admin_required] = script_data[:admin_required] || false # default false
+      property_hash[:professional_learning_course] = script_data[:professional_learning_course] || false # default false
+      property_hash[:student_of_admin_required] = script_data[:student_of_admin_required] || false # default false
+      property_hash[:peer_reviews_to_complete] = script_data[:peer_reviews_to_complete] if script_data[:peer_reviews_to_complete]
+    end
+  end
```
Comment (joshlory) [inline dashboard/app/models/script.rb]: Same feedback as above: https://github.com/code-dot-org/code-dot-org/pull/9293/files#r69778003

File: dashboard/app/models/script.rb [hunk 10]
```diff
@@ -608,11 +598,37 @@ def csf_finish_url
...
 
   def summarize
+    summarized_stages = stages.map(&:summarize)
+
+    if peer_reviews_to_complete
+      levels = []
+      peer_reviews_to_complete.times do |x|
+        levels << {
+            ids: [x],
+            position: 0,
```
Comment (joshlory) [inline dashboard/app/models/script.rb]: Do we use `position` anywhere when rendering progress dots?  Can it also be removed in `ScriptLevel#summarize`?

File: dashboard/app/helpers/users_helper.rb [hunk 11]
```diff
@@ -126,6 +127,19 @@ def summarize_trophies(script, user = current_user)
...
 
+  def merge_peer_review_progress(user_data, user, script)
+    if user &&
+        script.professional_learning_course? &&
+        Plc::EnrollmentUnitAssignment.exists?(user: user, plc_course_unit: script.plc_course_unit)
+
+      user_data[:peerReviewsPerformed] = PeerReview.where(reviewer: user, script: script).map do |review|
+        review.summarize.merge!(url: peer_review_path(review))
+      end
+    end
```
Comment (joshlory) [inline dashboard/app/helpers/users_helper.rb]: Can this logic be pushed down to the `PeerReview` model, and tested?

**Comment by joshlory** [comment]
A few spec-level suggestions for peer review:

screen shot 2016-07-06 at 11 36 26 am
- An un-started review is 'attempted' (yellow), should be 'not_started' (not filled in).
- I think it should be "Review unavailable"?
- Prefer the dots as a horizontal list, since they have an implied completion order.

peer-review-assignments

File: code-studio/src/js/components/progress/course_progress.jsx (hunk 1 above)
Comment (mehalshah) [inline code-studio/src/js/components/progress/course_progress.jsx]: We should use trailing commas. Makes it easier to delete / rearrange stuff.

Comment (joshlory) [inline code-studio/src/js/components/progress/course_progress.jsx]: Cool, I don't have any objections.  Let's enable the eslint rule to enforce this consistently!

Comment (joshlory) [inline code-studio/src/js/components/progress/course_progress.jsx]: Cool, no strong preference.  If we go with trailing commas let's enable the eslint rule to enforce it everywhere (separate PR).

File: code-studio/src/js/components/progress/progress_dot.jsx (hunk 3 above)
Comment (mehalshah) [inline code-studio/src/js/components/progress/progress_dot.jsx]: Right now, the only circumstance where a level would be locked is if its a locked peer review slot.

I can add a property the levels on whether or not a certain line is locked. Will try that out

File: code-studio/src/js/components/progress/progress_dot.jsx (hunk 4 above)
Comment (mehalshah) [inline code-studio/src/js/components/progress/progress_dot.jsx]: What's the better way to do this? Because the interior is the same whether the outer element is a div or an anchor.

File: dashboard/app/helpers/users_helper.rb (hunk 11 above)
Comment (mehalshah) [inline dashboard/app/helpers/users_helper.rb]: I can push this to the PeerReview model as a static method

File: dashboard/app/models/peer_review.rb (hunk 8 above)
Comment (mehalshah) [inline dashboard/app/models/peer_review.rb]: I was doing this with tap because I didn't want to set the icon field for in progress reviews. But that's no longer the case, so I'll do this method

File: dashboard/app/models/script.rb (hunk 9 above)
Comment (mehalshah) [inline dashboard/app/models/script.rb]: So here, I'm using tap because I don't want to set the peer_reviews_to_complete key for scripts that don't have peer reviews.

File: code-studio/src/js/components/progress/progress_dot.jsx (hunk 4 above)
Comment (joshlory) [inline code-studio/src/js/components/progress/progress_dot.jsx]: We chatted in person about keeping the wrapping `<a>` tag and removing the `href` attribute so it's not clickable.

File: dashboard/app/models/script.rb (hunk 9 above)
Comment (joshlory) [inline dashboard/app/models/script.rb]: Ah I see.  One option is to `.compact` the object to remove nil keys in Rails:

``` ruby
irb(main):001:0> {a: nil, b: 2}.compact
=> {:b=>2}
```

File: code-studio/src/js/components/progress/progress_dot.jsx (hunk 4 above)
Comment (mehalshah) [inline code-studio/src/js/components/progress/progress_dot.jsx]: Yeah - I tried it out earlier. I didn't like how it was automatically inheriting bootstrap anchor tag styling, but I think I can suppress it now.

File: dashboard/app/models/script.rb (hunk 9 above)
Comment (mehalshah) [inline dashboard/app/models/script.rb]: Oh that's fine

**Comment by mehalshah** [comment]
Tagging @ryansloan for comment on above

**Comment by ryansloan** [comment]
Re: the Horizontal List, I like it more from a visual standpoint (especially because it eliminates the repeated strings), but I worry that it could be a bit ambiguous. It's not necessarily clear why a given peer review is locked. Now I'm inventing features, but maybe we could do the row of dots with the status in a tooltip when you hover/click on a locked level?

**Comment by mehalshah** [comment]
I propose leaving it as in, then taking feedback from our bug bash next week and focus group the week after.

I'd like to better understand if people plan on coming back to peer reviews they've done, what order they go in when doing them, and what's clear / not clear about expectations for this section.

File: dashboard/app/models/script.rb (hunk 10 above)
Comment (mehalshah) [inline dashboard/app/models/script.rb]: We don't use it with dots, but I don't know where else we are using it right now. I'll remove it here but won't touch it in scriptlevel until I've understood that part more

**Comment by mehalshah** [comment]
Implemented most of the suggestions here - progress_dot.jsx isn't as messy as I thought it would be

File: dashboard/app/models/script.rb [hunk 12]
```diff
@@ -605,12 +605,12 @@ def summarize
       peer_reviews_to_complete.times do |x|
         levels << {
             ids: [x],
-            position: 0,
             kind: 'peer_review',
             title: '',
             url: '',
             name: I18n.t('peer_review.reviews_unavailable'),
-            icon: 'fa-lock'
+            icon: 'fa-lock',
```
Comment (joshlory) [inline dashboard/app/models/script.rb:612]: Let's have `locked` imply the `'fa-lock'` icon so this doesn't need to be passed separately.

File: code-studio/src/js/components/progress/progress_dot.jsx [hunk 13]
```diff
@@ -163,72 +167,55 @@ export const ProgressDot = React.createClass({
...
-      </div>;
-
-    const rowText = (showLevelName || isPeerReview) &&
-      <span
-        key='named_level'
-        style={[styles.levelName, isLockedReview && {color: color.charcoal}]}
+    return (
+      <a
+        key='link'
+        href={level.locked ? undefined : level.url}
```
Comment (joshlory) [inline code-studio/src/js/components/progress/progress_dot.jsx]: Nit: `href={level.locked && level.url}`

File: code-studio/src/js/components/progress/progress_dot.jsx [hunk 14]
```diff
@@ -163,72 +167,55 @@ export const ProgressDot = React.createClass({
...
+        onClick={this.props.saveAnswersBeforeNavigation && dotClicked.bind(null, level.url)}
+        style={[styles.outer, (showLevelName || isPeerReview) && {display: 'table-row'}, level.locked && styles.disabledLevel]}
```
Comment (joshlory) [inline code-studio/src/js/components/progress/progress_dot.jsx]: Nit: break these out into separate lines when they exceed the 80 char limit.

``` jsx
style={[
  styles.outer,
  (showLevelName || isPeerReview) && {display: 'table-row'},
  level.locked && styles.disabledLevel
]}
```

File: code-studio/src/js/components/progress/progress_dot.jsx [hunk 15]
```diff
@@ -163,72 +167,55 @@ export const ProgressDot = React.createClass({
...
       >
-        {level.name}
-      </span>;
-
-
-    if (isLockedReview) {
-      return (
+        {(level.icon && !isPeerReview) ?
+        <i
```
Comment (joshlory) [inline code-studio/src/js/components/progress/progress_dot.jsx]: This continuation needs indent:

``` jsx
{level.icon && !isPeerReview ?
  <i
    ...
  /> :
  <div
    ...
  >
    ...
  </div>
}
```

File: code-studio/src/js/components/progress/progress_dot.jsx [hunk 16]
```diff
@@ -163,72 +167,55 @@ export const ProgressDot = React.createClass({
...
+            styles.dot.icon,
+            smallDot && styles.dot.icon_small,
+            level.status && level.status !== 'not_tried' && styles.dot.icon_complete,
+            outlineCurrent && {textShadow: createOutline(color.level_current)}
+          ]}
+        /> :
         <div
-          key='lockedReview'
-          style={[styles.outer, {display: 'table-row'}]}
+          className={`level-${level.id} ${isPeerReview && `fa ${level.icon}`}`}
```
Comment (joshlory) [inline code-studio/src/js/components/progress/progress_dot.jsx]: `${isPeerReview && ...}` can go away when https://github.com/code-dot-org/code-dot-org/pull/9293/commits/6cb1aa617ccfd661cffff07b46015d6ab5c0fdfd#r70003929 is addressed.

File: code-studio/src/js/components/progress/progress_dot.jsx [hunk 17]
```diff
@@ -163,72 +167,55 @@ export const ProgressDot = React.createClass({
...
-      return (
-        <a
-          key='link'
-          href={level.url}
-          onClick={this.props.saveAnswersBeforeNavigation && dotClicked.bind(null, level.url)}
-          style={[styles.outer, (showLevelName || isPeerReview) && {display: 'table-row'}]}
+          {(showLevelName || (isPeerReview && level.icon === '')) ? '\u00a0' : level.title}
+        </div>}
+        {(showLevelName || isPeerReview) &&
+        <span
```
Comment (joshlory) [inline code-studio/src/js/components/progress/progress_dot.jsx]: Same continuation indent issue.

File: code-studio/src/js/components/progress/progress_dot.jsx [hunk 18]
```diff
@@ -156,49 +166,56 @@ export const ProgressDot = React.createClass({
...
+    const isPeerReview = level.kind === 'peer_review';
 
     return (
       <a
         key='link'
-        href={level.url}
+        href={level.locked ? undefined : level.url}
         onClick={this.props.saveAnswersBeforeNavigation && dotClicked.bind(null, level.url)}
-        style={[styles.outer, showLevelName && {display: 'table-row'}]}
+        style={[styles.outer, (showLevelName || isPeerReview) && {display: 'table-row'}, level.locked && styles.disabledLevel]}
```
Comment (joshlory) [inline code-studio/src/js/components/progress/progress_dot.jsx]: Can `showLevelName` be set to `/(named_level|peer_review)/.test(level.kind)  && this.props.courseOverviewPage` to clean up all these `|| isPeerReview`?

**Comment by joshlory** [comment]
Still seeing "Review in progress" for reviews I haven't started yet:

screen shot 2016-07-07 at 4 28 56 pm

**Edit:** this definitely shouldn't block check-in, can be part of a future PR.

File: code-studio/src/js/components/progress/progress_dot.jsx [hunk 19]
```diff
@@ -156,49 +166,56 @@ export const ProgressDot = React.createClass({
...
+        >
+          {level.name}
+        </span>}
       </a>
     );
   }
 });
-export default connect(state => ({
+
+export default connect((state) => ({
```
Comment (joshlory) [inline code-studio/src/js/components/progress/progress_dot.jsx]: Super small nit: extra parens diff isn't needed.

File: code-studio/src/js/components/progress/types.js [hunk 20]
```diff
@@ -4,10 +4,11 @@ const levelProgressShape = React.PropTypes.shape({
   title: React.PropTypes.oneOfType([React.PropTypes.string, React.PropTypes.number]),
   name: React.PropTypes.string,
   status: React.PropTypes.string,
-  kind: React.PropTypes.oneOf(['unplugged', 'assessment', 'puzzle', 'named_level']),
+  kind: React.PropTypes.oneOf(['unplugged', 'assessment', 'puzzle', 'named_level', 'peer_review']),
   icon: React.PropTypes.string,
   url: React.PropTypes.string,
-  id: React.PropTypes.oneOfType([React.PropTypes.string, React.PropTypes.number])
+  id: React.PropTypes.oneOfType([React.PropTypes.string, React.PropTypes.number]),
+  locked: React.PropTypes.bool
```
Comment (joshlory) [inline code-studio/src/js/components/progress/types.js]: Nit: trailing comma?  (One more below.)

File: code-studio/src/js/progress.js [hunk 21]
```diff
@@ -157,14 +158,33 @@ function loadProgress(scriptData, currentLevelId, saveAnswersBeforeNavigation =
...
 
-          return Object.assign({}, level, {status: progress.activityCssClass(newProgress[id])});
-        })}))
+      const peerReviewStage = _.findLast(returnState.stages, function (stage) {
+        return stage.flex_category === 'Peer Review';
       });
+
+      if (action.peerReviewsPerformed) {
+        action.peerReviewsPerformed.forEach(function (peerReview, index) {
+            Object.assign(peerReviewStage.levels[index], peerReview);
```
Comment (joshlory) [inline code-studio/src/js/progress.js]: Is `peerReviewStage` unused?

Comment (joshlory) [inline code-studio/src/js/progress.js]: Never mind, I see.

Comment (joshlory) [inline code-studio/src/js/progress.js]: This is a little confusing as written.  Can you move this logic into `state.stages.map` above, inside an `if (stage.flex_category === 'Peer Review')`?

File: dashboard/app/models/script.rb (hunk 12 above)
Comment (mehalshah) [inline dashboard/app/models/script.rb:612]: I'm not sure...I guess it depends on what we'd want to happen if we locked a level that had an icon.

File: code-studio/src/js/components/progress/progress_dot.jsx (hunk 13 above)
Comment (joshlory) [inline code-studio/src/js/components/progress/progress_dot.jsx]: Never mind, how you have it written is better.

File: dashboard/app/helpers/users_helper.rb [hunk 22]
```diff
@@ -9,6 +9,17 @@ def summarize_user_progress(script, user = current_user, exclude_level_progress
...
+
+    peer_review_summaries = PeerReview.get_peer_review_summaries(user, script)
+
+    unless peer_review_summaries.empty?
+      peer_review_summaries.each do |summary|
+        summary.merge!(url: peer_review_path(summary[:id]))
+      end
+
+      user_data[:peerReviewsPerformed] = peer_review_summaries
+    end
```
Comment (joshlory) [inline dashboard/app/helpers/users_helper.rb:15]: Simplify to this?

``` ruby
user_data[:peerReviewsPerformed] = PeerReview.get_peer_review_summaries(user, script).map do |summary|
  summary.merge(url: peer_review_path(summary[:id]))
end
```

(Can return `user_data.compact` if you're worried about the possible nil value.)

File: code-studio/src/js/components/progress/progress_dot.jsx (hunk 16 above)
Comment (mehalshah) [inline code-studio/src/js/components/progress/progress_dot.jsx]: I don't see how this goes away - it's needed to be fa-check for a completed review, fa-lock for a locked row, and nothing in other cases.

File: dashboard/app/models/peer_review.rb [hunk 23]
```diff
@@ -133,4 +139,28 @@ def self.create_for_submission(user_level, level_source_id)
...
   end
+
+  def self.get_peer_review_summaries(user, script)
+    peer_review_summaries = []
+
+    if user &&
+        script.professional_learning_course? &&
+        Plc::EnrollmentUnitAssignment.exists?(user: user, plc_course_unit: script.plc_course_unit)
+
+      peer_review_summaries = PeerReview.where(reviewer: user, script: script).map(&:summarize)
```
Comment (joshlory) [inline dashboard/app/models/peer_review.rb]: I don't think `peer_review_summaries` is needed — can just `return PeerReview.where...` here.

File: dashboard/app/models/script.rb (hunk 12 above)
Comment (joshlory) [inline dashboard/app/models/script.rb:612]: I guess that's possible, but in that case I think we'd still want to show the default locked styling (so that the level being locked is more important that what type of level it is).

File: dashboard/app/models/peer_review.rb [hunk 24]
```diff
@@ -140,13 +140,27 @@ def self.create_for_submission(user_level, level_source_id)
...
+    peer_review_summaries
+  end
+
+  def summarize
+    return {
+      id: id,
+      status: status.nil? ? 'not_started' : 'perfect',
+      name: status.nil? ? I18n.t('peer_review.review_in_progress') : I18n.t('peer_review.link_to_submitted_review'),
+      result: status.nil? ? ActivityConstants::UNSUBMITTED_RESULT : ActivityConstants::BEST_PASS_RESULT,
+      icon: status.nil? ? '' : 'fa-check',
```
Comment (joshlory) [inline dashboard/app/models/peer_review.rb]: This can go away when https://github.com/code-dot-org/code-dot-org/pull/9293/commits/6cb1aa617ccfd661cffff07b46015d6ab5c0fdfd#r70003929 is addressed — completed peer reviews should have the same styling as completed tasks.

File: code-studio/src/js/progress.js (hunk 21 above)
Comment (mehalshah) [inline code-studio/src/js/progress.js]: I broke it out this way because stages are composed of levels, but the peer review section is composed of reviews. The above map statement was already hard to parse (ending with })})); ) and I didn't see a good place to insert something that wasn't mapping levels within a stage.

File: code-studio/src/js/components/progress/progress_dot.jsx (hunk 16 above)
Comment (mehalshah) [inline code-studio/src/js/components/progress/progress_dot.jsx]: `className={`level-${level.id} fa ${level.locked ? 'fa-lock' : level.icon}`}`
doesn't look much cleaner, not to mention the line with the \u00a0 needs to know if an icon was rendered

File: code-studio/src/js/progress.js (hunk 21 above)
Comment (joshlory) [inline code-studio/src/js/progress.js]: What about this?

``` javascript
if (action.type === 'MERGE_PROGRESS') {
  // TODO: _.mergeWith after upgrading to Lodash 4+
  let newProgress = {};
  Object.keys(Object.assign({}, state.progress, action.progress)).forEach(key => {
    newProgress[key] = clientState.mergeActivityResult(state.progress[key], action.progress[key]);
  });

  const stages = state.stages.map(stage => Object.assign({}, stage, {levels: stage.levels.map((level, index) => {
    const id = level.uid || progress.bestResultLevelId(level.ids, newProgress);

    if (action.peerReviewsPerformed && stage.flex_category === 'Peer Review') {
      Object.assign(level, action.peerReviewsPerformed[index]);
    }

    return Object.assign({}, level, {
      status: progress.activityCssClass(newProgress[id]),
      id: id,
      url: level.url
    });
  })}));

  return Object.assign({}, state, {
    progress: newProgress,
    stages: stages
  });
}
```

Comment (joshlory) [inline code-studio/src/js/progress.js]: My reason for pushing back here is that this reducer is already super complex, and I'd like to keep it as readable as possible.

File: code-studio/src/js/components/progress/progress_dot.jsx (hunk 16 above)
Comment (mehalshah) [inline code-studio/src/js/components/progress/progress_dot.jsx]: Eh...actually it kind of does

**Comment by joshlory** [comment]
This is looking awesome!  My main feedback this pass is to make the peer review dot styling as close as possible to task & make `locked` generic to any dot type.  Happy to chat about any of these in person.

File: code-studio/src/js/components/progress/progress_dot.jsx (hunk 16 above)
Comment (joshlory) [inline code-studio/src/js/components/progress/progress_dot.jsx]: Why do completed peer reviews have a check mark, when completed tasks do not?

File: code-studio/src/js/progress.js (hunk 21 above)
Comment (mehalshah) [inline code-studio/src/js/progress.js]: Okay - but `status: progress.activityCssClass(newProgress[id])` needs to not get assigned for peer reviews.

File: dashboard/app/models/peer_review.rb (hunk 24 above)
Comment (mehalshah) [inline dashboard/app/models/peer_review.rb]: Spec has them showing a check mark, not a solid green bubble.

@ryansloan ?

File: dashboard/app/models/peer_review.rb (hunk 23 above)
Comment (mehalshah) [inline dashboard/app/models/peer_review.rb]: I'd like to return empty array here so I've made it

```
if user &&
        script.professional_learning_course? &&
        Plc::EnrollmentUnitAssignment.exists?(user: user, plc_course_unit: script.plc_course_unit)

      PeerReview.where(reviewer: user, script: script).map(&:summarize)
    else
      []
    end
```

File: dashboard/app/models/peer_review.rb (hunk 24 above)
Comment (ryansloan) [inline dashboard/app/models/peer_review.rb]: Mockups reflect a check-mark in bubble for every task type (including completed tasks and peer reviews) in PL scripts: https://docs.google.com/document/d/1Nk8n3Kdir_qTEeR44QdOGgNUskhErSaEqf94YwsuBfs/edit#

Comment (mehalshah) [inline dashboard/app/models/peer_review.rb]: Okay - will fix

**Comment by mehalshah** [comment]
Oh bother, this implementation doesn't work anyway

image

File: code-studio/src/js/progress.js [hunk 25]
```diff
@@ -157,13 +158,23 @@ function loadProgress(scriptData, currentLevelId, saveAnswersBeforeNavigation =
...
+        if (action.peerReviewsPerformed && stage.flex_category === 'Peer Review') {
+          Object.assign(level, action.peerReviewsPerformed[index]);
+        }
+
+        return Object.assign({}, level, {
+          status: level.kind === 'peer_review' ? level.status : progress.activityCssClass(newProgress[id]),
+          id: id,
+          url: level.url
+        });
+      })}));
```
Comment (Bjvanminnen) [inline code-studio/src/js/progress.js]: FYI, I have an open PR that might result in some merge conflicts with this. https://github.com/code-dot-org/code-dot-org/pull/9379

File: code-studio/src/js/components/progress/progress_dot.jsx [hunk 26]
```diff
@@ -138,12 +138,40 @@ const styles = {
   }
 };
 
-
 function dotClicked(url, e) {
   e.preventDefault();
   saveAnswersAndNavigate(url);
 }
 
+const BubbleInterior = React.createClass({
```
Comment (joshlory) [inline code-studio/src/js/components/progress/progress_dot.jsx]: Let's pull this out to a separate file.

Comment (Bjvanminnen) [inline code-studio/src/js/components/progress/progress_dot.jsx]: I had actually recommended leaving it in the file. It's a pretty simple component that's not used anywhere else. (Could also be implemented as a stateless functional component).

**Comment by joshlory** [comment]
The test coverage you've written looks good, but it would be great to have a UI test to cover this scenario end-to-end.  Doesn't need to block this PR.

**Comment by mehalshah** [comment]
My priority right now is getting something that the team can bash on Thursday, and that I can give to facilitators next Wednesday. After that, I'd like to make an eyes test to verify the appearance of unit view for someone who's partially done with stuff

File: code-studio/src/js/components/progress/progress_dot.jsx (hunk 26 above)
Comment (joshlory) [inline code-studio/src/js/components/progress/progress_dot.jsx]: Ah, makes sense.  I'm used to one-class-per-file to make it easy to find things, but I can see the exception when it's only used one place.

**Comment by joshlory** [comment]
Hmm I'm getting the following after completing a peer review: `Warning: Failed propType: Invalid prop `title` of type `string` supplied to `BubbleInterior`, expected `number`. Check the render method of `ProgressDot`.`

File: code-studio/src/js/components/progress/progress_dot.jsx (hunk 26 above)
Comment (Bjvanminnen) [inline code-studio/src/js/components/progress/progress_dot.jsx]: Yeah, I think one-class-per-file should be the general rule. I think to qualify for an exception the following needs to hold true.
1. Component is used in only this file
2. Component has a reasonably simple render method, and no other methods.

**Comment by joshlory** [comment]
LGTM!  (After resolving `rake test` failure.)
